libraries/anvill_passes/src/TransformRemillJumpIntrinsics.cpp

```diff
@@ -0,0 +1,229 @@
+ReturnAddressResult
+TransformRemillJumpIntrinsics::QueryReturnAddress(const llvm::DataLayout &dl,
+                                                  llvm::Value *val) const {
+
+  if (auto call = llvm::dyn_cast<llvm::CallBase>(val)) {
+    if (call->getIntrinsicID() == llvm::Intrinsic::returnaddress) {
+      return kReturnAddressProgramCounter;
+    } else if (auto func = call->getCalledFunction()) {
+      if (func->getName().startswith("__remill_read_memory_")) {
+        auto addr = call->getArgOperand(1);  // Address
+
+        // Could it be address of return address ??
+        auto addr_of_ret = llvm::dyn_cast<llvm::CallBase>(addr);
+        if (addr_of_ret->getIntrinsicID() ==
+            llvm::Intrinsic::addressofreturnaddress) {
+          return kReturnAddressProgramCounter;
+        }
+      }
+    }
+
+  } else if (auto gv = llvm::dyn_cast<llvm::GlobalVariable>(val);
+             gv && IsReturnAddress(gv)) {
+    return kReturnAddressProgramCounter;
+
+  } else if (auto pti = llvm::dyn_cast<llvm::PtrToIntOperator>(val)) {
+    return QueryReturnAddress(dl, pti->getOperand(0));
+
+  // Sometimes optimizations result in really crazy looking constant expressions
+  // related to `__anvill_ra`, full of shifts, zexts, etc. We try to detect
+  // this situation by initializing a "magic" address associated with
+  // `__anvill_ra`, and then if we find this magic value on something that
+  // references `__anvill_ra`, then we conclude that all those manipulations
+  // in the constant expression are actually not important.
+  } else if (auto xr = xref_resolver_.TryResolveReference(val);
+             xr.is_valid && xr.references_return_address &&
+             xr.u.address == xref_resolver_.MagicReturnAddressValue()) {
+    return kReturnAddressProgramCounter;
+  }
+
+  return kUnclassifiableProgramCounter;
+}
+
+// Find the remill intrinsic in module. If it is missing create one
+// with the given function type
+static llvm::Function *FindIntrinsic(llvm::Module *module,
+                                     llvm::FunctionType *type,
+                                     const char *name) {
+  auto function = module->getFunction(name);
+  if (!function) {
+    function = llvm::Function::Create(type, llvm::GlobalValue::ExternalLinkage,
+                                      name, module);
+  }
+
+  CHECK(nullptr != function) << "Unable to get intrinsic: " << name;
+  function->addFnAttr(llvm::Attribute::NoDuplicate);
+  function->addFnAttr(llvm::Attribute::NoUnwind);
+  function->addFnAttr(llvm::Attribute::OptimizeNone);
+  function->addFnAttr(llvm::Attribute::NoInline);
+  function->removeFnAttr(llvm::Attribute::NoReturn);
+  function->removeFnAttr(llvm::Attribute::UWTable);
+  function->removeFnAttr(llvm::Attribute::AlwaysInline);
+  return function;
+}
+
+// Find the call site of the given function and add them to vector
+// if `pred(call)` is true
+static std::vector<llvm::CallBase *>
```

Comment:
Given that this is `static`, instead of using `std::function`, use a template typename, so that the compiler can inline the predicate lambda into the static function. `std::function` often hurts optimization.